packages/connector-oauth2/src/index.ts

```diff
@@ -0,0 +1,141 @@
+import type {
+  GetAuthorizationUri,
+  GetUserInfo,
+  SocialConnector,
+  CreateConnector,
+  GetConnectorConfig,
+} from '@logto/connector-kit';
+import {
+  ConnectorError,
+  ConnectorErrorCodes,
+  validateConfig,
+  ConnectorType,
+  parseJson,
+} from '@logto/connector-kit';
+import got, { HTTPError } from 'got';
+import omit from 'lodash.omit';
+import snakecaseKeys from 'snakecase-keys';
+
+import { defaultMetadata, defaultTimeout, oauthConfigGlobalKeys } from './constant';
+import type { OauthConfig } from './types';
+import {
+  authResponseGuard,
+  implicitAuthResponseGuard,
+  oauthConfigGuard,
+  OauthGrantType,
+} from './types';
+import { accessTokenRequester, userProfileMapping } from './utils';
+
+const getAuthorizationUri =
+  (getConfig: GetConnectorConfig): GetAuthorizationUri =>
+  async ({ state, redirectUri }) => {
+    const config = await getConfig(defaultMetadata.id);
+    validateConfig<OauthConfig>(config, oauthConfigGuard);
+    const parsedConfig = oauthConfigGuard.parse(config);
+
+    if (
+      parsedConfig.oauthGrantType === OauthGrantType.AuthorizationCode ||
+      parsedConfig.oauthGrantType === OauthGrantType.Implicit
+    ) {
+      const { redirect_uri: configured_redirect_uri, ...rest } = snakecaseKeys(
+        omit(parsedConfig, ...oauthConfigGlobalKeys, 'clientSecret', 'grantType')
+      );
+
+      const queryParameters = new URLSearchParams({
+        state,
+        /**
+         * TODO: GetAuthorizationUri should have optional redirect_uri or truncate redirect_uri.
+         * Maybe should make redirect_uri optional to be compatible to previous versions.
+         */
+        ...rest,
+        redirect_uri: configured_redirect_uri ?? redirectUri,
+      });
+
+      return `${parsedConfig.authorizationEndpoint}?${queryParameters.toString()}`;
+    }
+
+    // TODO: some flows do not access authorization endpoint, should fix API
```

Comment:
Then better throw an error?

Reply:
We call `getAuthorizationUri` and then `getUserInfo` for all social connectors, hence we can not simply throw an error here. Need to have at least two handlers when we get different authorization URIs.